File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/sealedTypes/deprecation/deprecation.kt
// KIND: STANDALONE
// MODULE: SealedDeprecation
// FILE: deprecation.kt

// A warning-level inheritor keeps its case; an error-level one is unavailable in Swift and falls
// into `unknown`; a deprecated root still reports its inheritors.

package org.kotlin.foo

sealed class SealedClassNonDeprecated

@Deprecated("unavailable", level = DeprecationLevel.ERROR)
class DeprecatedErrorSubClass : SealedClassNonDeprecated() {
    override fun toString(): String = "DeprecatedErrorSubClass"
}

@Deprecated("deprecated")
class DeprecatedWarningSubClass : SealedClassNonDeprecated() {
    override fun toString(): String = "DeprecatedWarningSubClass"
}

// An error-level deprecated root: unavailable in Swift, so nothing here is referenceable from the
// test; the enum ends up with nothing but an `unknown` case.
@Deprecated("unavailable", level = DeprecationLevel.ERROR)
sealed class SealedClassDeprecatedError

@Suppress("DEPRECATION_ERROR")
class NonDeprecatedSubClassA : SealedClassDeprecatedError()

@Deprecated("deprecated")
sealed class SealedClassDeprecatedWarning

class NonDeprecatedSubClassB : SealedClassDeprecatedWarning() {
    override fun toString(): String = "NonDeprecatedSubClassB"
}

// FILE: factories.kt
package org.kotlin.foo

@Suppress("DEPRECATION")
fun createDeprecatedWarningSubClass(): SealedClassNonDeprecated = DeprecatedWarningSubClass()

@Suppress("DEPRECATION_ERROR")
fun createDeprecatedErrorSubClass(): SealedClassNonDeprecated = DeprecatedErrorSubClass()

@Suppress("DEPRECATION")
fun createNonDeprecatedSubClassB(): SealedClassDeprecatedWarning = NonDeprecatedSubClassB()

File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/sealedTypes/deprecation/deprecation.swift
import KotlinRuntime
import SealedDeprecation
import Testing

private typealias SealedClassNonDeprecated = ExportedKotlinPackages.org.kotlin.foo.SealedClassNonDeprecated

// Naming the deprecated declarations warns on purpose. The `@available` attributes themselves have
// no runtime effect, so what this case covers is which cases exist, not how they are annotated.

private func describe(_ value: SealedClassNonDeprecated) -> String {
    switch value.sealedType() {
        case let .deprecatedWarningSubClass(type): "warning: \(type.value)"
        // `DeprecatedErrorSubClass` is `@available(*, unavailable)`, so it has no case at all.
        case let .unknown(type): "unknown: \(type.value)"
    }
}

@Test
func testWarningDeprecatedInheritorKeepsItsOwnCase() throws {
    let value = ExportedKotlinPackages.org.kotlin.foo.createDeprecatedWarningSubClass()
    try #require(describe(value) == "warning: DeprecatedWarningSubClass")
}

@Test
func testErrorDeprecatedInheritorFallsBackToUnknown() throws {
    let value = ExportedKotlinPackages.org.kotlin.foo.createDeprecatedErrorSubClass()
    try #require(describe(value) == "unknown: DeprecatedErrorSubClass")
}

@Test
func testDeprecatedRootStillReportsItsInheritor() throws {
    let value = ExportedKotlinPackages.org.kotlin.foo.createNonDeprecatedSubClassB()
    let result = switch value.sealedType() {
        case let .nonDeprecatedSubClassB(type): "inheritor: \(type.value)"
    }
    try #require(result == "inheritor: NonDeprecatedSubClassB")
}
